DRILL-5356: Refactor Parquet Record Reader

paul-rogers opened the pull request:
The Parquet reader is Drill's premier data source and has worked very well
for many years. As with any piece of code, it has grown in complexity over
that time and has become hard to understand and maintain.

In work in another project, we found that Parquet is accidentally creating
"low density" batches: record batches with little actual data compared to
the amount of memory allocated. We'd like to fix that.

However, the current complexity of the reader code creates a barrier to
making improvements: the code is so complex that it is often better to
leave bugs unfixed, or risk spending large amounts of time struggling to
make small changes.

This commit offers to help revitalize the Parquet reader. Functionality is
identical to the code in master; but code has been pulled apart into
various classes each of which focuses on one part of the task: building
up a schema, keeping track of read state, a strategy for reading various
combinations of records, etc. The idea is that it is easier to understand
several small, focused classes than one huge, complex class. Indeed, the
idea of small, focused classes is common in the industry; it is nothing new.

Unit tests pass with the change. Since no logic has chanaged, we only moved
lines of code, that is a good indication that everything still works.

ppadma commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/ReadState.java:
```diff
@@ -0,0 +1,157 @@
+  }
+
+  public ParquetSchema schema() { return schema; }
+  public List<ColumnReader<?>> getReaders() { return columnStatuses; }
```
for clarity, should we rename this function getColumnReaders ?

ppadma commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/BatchReader.java:
```diff
@@ -0,0 +1,164 @@
+    @Override
+    protected long getReadCount(ColumnReader<?> firstColumnStatus) {
+      if (readState.mockRecordsRead == readState.schema().getGroupRecordCount()) {
+        return 0;
```
How about moving mockRecordsRead to this class instead of keeping it in readState ?

ppadma commented on exec/java-exec/src/test/java/org/apache/drill/exec/store/parquet/ParquetInternalsTest.java:
```diff
@@ -0,0 +1,161 @@
+      .run();
+  }
+
+
```
remove extra line

paul-rogers replied:
Fixed.

ppadma commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/ParquetSchema.java:
```diff
@@ -0,0 +1,262 @@
+    }
+    return false;
+  }
+  /**
```
add a new line after end of function

paul-rogers replied:
Fixed.

paul-rogers commented:
Squashed and rebased on master. @parthchandra, Padma would like you to review this in addition to her review.

This PR would be incredibly helpful for the next step in the project to fix memory fragmentation: we will implement code to limit batch size -- kind of like the original intent, but this time it should work. That work will be much easier to do on top of the refactored code than on top of the original code.

parthchandra commented:
+1 LGTM.  Thanks for the cleanup Paul!

sudheeshkatkam commented:
Are the changes only in 1494915dbef5dbd5996c19d0a2e89ca450a8ae3a (to cherry pick)?

parthchandra commented:
I took the entire patch and applied it to master (use git am -3). Git manages to figure out that the commits are already applied. One commit caused a merge conflict and I skipped it. In the end it left me with only the one commit.

paul-rogers commented:
Thanks. I'll clean up the messy commits today. Not sure how it picked up the other six commits...

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/BatchReader.java:
```diff
@@ -0,0 +1,164 @@
+
+  /**
+   * Strategy for reading mock records. (What are these?)
+   */
```
Fixed. Finally found out what this means. Thanks Jinfeng!

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/ParquetColumnMetadata.java:
```diff
@@ -0,0 +1,151 @@
+   */
+  private int getDataTypeLength() {
+    if (! isFixedLength()) {
+      return -1;
```
Original code, but sure, since I'm mucking about, I defined UNDEFINED_LENGTH.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/ParquetColumnMetadata.java:
```diff
@@ -0,0 +1,151 @@
+  }
+
+  @SuppressWarnings("resource") FixedWidthRepeatedReader makeRepeatedFixedWidthReader(ParquetRecordReader reader, int recordsPerBatch) throws Exception {
+    final RepeatedValueVector repeatedVector = RepeatedValueVector.class.cast(vector);
```
Fixed.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/ParquetRecordReader.java:
```diff
@@ -308,163 +232,50 @@ public FragmentContext getFragmentContext() {
-          }
-        }
-      }
+      schema.buildSchema(footer, batchSize);
```
Nice improvement. Thanks!

paul-rogers commented:
Cleaned up the multi-commit mess, rebased on the latest master, and fixed minor issues raised in code review comments. Should be read to commit.
